### DISCUSSION TITLE
What is the difference between onOperation and onSubscribe hooks ?

### DISCUSSION THREAD

**Discussion by 97amarnathk** [enisdenjo/graphql-ws#161] [Q&A] [answered]
Hey, I was going through the graphql-ws docs, and found that here we have 2 different hooks:
* `onOperation`
* `onSubscribe`

My understanding is that when I make a new graphql subscription, the hooks are executed in the following order:
* onConnect
* onOperation
* onSubscribe
* for each payload sent from server to client:
    * onNext

Am I missing anything out in this ? 

So if this is the order, then in the `custom authentication` example, why is the auth being done in `onSubscribe`? Shouldnt it be done in `onOperation` ?

I previously used `subscriptions-transport-ws` and over there we did authentication inside `onOperation`. I am assuming that graphql-ws and subscription-transport-ws have the use for `onOperation`

**Comment by enisdenjo** [comment, marked as answer]
You might be mixing "subscribe" with GraphQL subscriptions, but that is not the case. The client "subscribes" for events on the server, `onSubscribe` is the first to be called. It allows you to manipulate the the execution arguments and skip the built-in validation and exec args construction. With it you can use dynamic schemas, implement persisted queries, skip validation, apply custom auth or introduce completely custom subscription behaviour (check the recipes out). Check it out here:
https://github.com/enisdenjo/graphql-ws/blob/f24e9bac25b705b53db68e6b71c193d13b4e324a/src/server.ts#L619-L654

`onOperation`, on the other hand, gets called immediately **after** the operation (query, mutation or subscription) is performed by GraphQL. It allows you to manipulate the execution results. Check it out here:
https://github.com/enisdenjo/graphql-ws/blob/f24e9bac25b705b53db68e6b71c193d13b4e324a/src/server.ts#L676-L690

**Reply by enisdenjo** [reply]
The complete order of callbacks is then:
1. [`onConnect`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#onconnect)
1. [`onSubscribe`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#onsubscribe)
1. [`onOperation`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#onoperation)
1. [`onNext`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#onnext)
1. [`onError`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#onerror) or [`onComplete`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#oncomplete)
1. [`onDisconnect`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#ondisconnect)
1. [`onClose`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#onclose)

**Reply by 97amarnathk** [reply]
Okay. this is helpful. Thanks! 👍🏽
